File: python_june_2024/17. Letter Combinations of a Phone Number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        mappings = {"2": ['a','b','c'],
                    "3": ['d','e','f'],
                    "4": ['g','h','i'],
                    "5": ['j','k','l'],
                    "6": ['m','n','o'],
                    "7": ['p','q','r','s'],
                    "8": ['t','u','v'],
                    "9": ['w','x','y','z']}
        old_ans = [""]
        for d in digits:
            new_ans = []
            for a in old_ans:
                for c in mappings[d]:
                    new_ans.append(a + c)
            old_ans = new_ans
        
        if old_ans == [""]:
            return []
        return old_ans

File: java_august/17. Letter Combinations of a Phone Number.java
class Solution {
    private List<String> result;
    private Map<Character, String> mapping = Map.of(
        '2',"abc",'3',"def",'4',"ghi",'5',"jkl",
        '6',"mno",'7',"pqrs",'8',"tuv",'9',"wxyz"
    );

    public List<String> letterCombinations(String digits) {
        result = new ArrayList<>();
        int n = digits.length();
        if (n==0) {
            return result;
        }
        backtrack(digits, 0, new StringBuilder(), n);
        return result;
    }

    private void backtrack(String digits, int idx, StringBuilder path, int n) {
        if (idx == n) {
            result.add(path.toString());
            return;
        }
        String chars = mapping.get(digits.charAt(idx));
        for (char c : chars.toCharArray()) {
            path.append(c);
            backtrack(digits, idx+1, path, n);
            path.deleteCharAt(path.length()-1);
        }
    }
}